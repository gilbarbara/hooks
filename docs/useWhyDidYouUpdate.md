# useWhyDidYouUpdate

A debugging hook that helps identify which props or state changes are causing a component to re-render.  
This is especially useful for optimizing performance and debugging unnecessary re-renders in React components.

## Usage

```tsx
function Component(props: { count: number; name: string }) {
  const changes = useWhyDidYouUpdate(props, 'MyComponent');

  return (
    <div>
      <p>Count: {props.count}</p>
      <p>Name: {props.name}</p>
    </div>
  );
}
```

**Skipping Logs and Inspecting Changes**

```tsx
import { useWhyDidYouUpdate } from '@gilbarbara/hooks';

function Component(props: { count: number; name: string }) {
  const changes = useWhyDidYouUpdate(props, { skipLog: true });

  return (
    <div>
      <p>Count: {props.count}</p>
      <p>Name: {props.name}</p>
      <pre>{JSON.stringify(changes, null, 2)}</pre>
    </div>
  );
}
```

## Reference

```typescript
interface UseWhyDidYouUpdateOptions {
  name?: string;
  skipLog?: boolean;
}

type UseWhyDidYouUpdateResult<T> = {
  [K in keyof T]?: {
    from: any;
    to: any;
  };
};

useWhyDidYouUpdate<T extends Record<string, any>>(
  props: T,
  nameOrOptions: string | UseWhyDidYouUpdateOptions = {},
): UseWhyDidYouUpdateResult<T>;
```
