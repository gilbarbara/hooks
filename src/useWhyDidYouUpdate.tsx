import { useEffect, useRef, useState } from 'react';

import { PlainObject } from './types';
import { isString } from './utils';

export type UseWhyDidYouUpdateResult<T> = {
  [K in keyof T]?: {
    from: any;
    to: any;
  };
};

export interface UseWhyDidYouUpdateOptions {
  name?: string;
  skipLog?: boolean;
}

export function useWhyDidYouUpdate<T extends PlainObject<any>>(
  props: T,
  nameOrOptions: string | UseWhyDidYouUpdateOptions = {},
): UseWhyDidYouUpdateResult<T> | false {
  type K = keyof T;

  const [changes, setChanges] = useState<UseWhyDidYouUpdateResult<T>>({});
  const previousProps = useRef<T>();

  const { name, skipLog = false } = isString(nameOrOptions)
    ? { name: nameOrOptions }
    : nameOrOptions;

  useEffect(() => {
    const { current } = previousProps;

    setChanges({});

    if (current) {
      // Get all keys from previous and current props
      const allKeys = Object.keys({ ...current, ...props }) as K[];

      // Use this object to keep track of changed props
      const changesObject: UseWhyDidYouUpdateResult<T> = {};

      // Iterate through keys
      allKeys.forEach(key => {
        if (current[key] !== props[key]) {
          changesObject[key] = {
            from: current[key],
            to: props[key],
          };
        }
      });

      // If changesObj not empty then output to console
      if (Object.keys(changesObject).length) {
        setChanges(changesObject);

        if (!skipLog) {
          const nameToken = name ? `: ${name}` : '';

          console.log(`[why-did-you-update${nameToken}]`, changesObject); // eslint-disable-line no-console
        }
      }
    }

    // Finally update previousProps with current props for next hook call
    previousProps.current = props;
  }, [name, props, skipLog]);

  return Object.keys(changes).length ? changes : false;
}
