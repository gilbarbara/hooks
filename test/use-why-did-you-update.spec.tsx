import * as React from 'react';
import { render } from '@testing-library/react';

import useWhyDidYouUpdate from '../src/use-why-did-you-update';

describe('useWhyDidYouUpdate', () => {
  const Component = (props: any) => {
    const changes = useWhyDidYouUpdate(props, { skipLog: true });

    return <div>{!!changes && JSON.stringify(changes, null, 2)}</div>;
  };

  const { container, rerender } = render(<Component ham={1} />);

  it('should show the changes', () => {
    expect(container.firstChild).toMatchSnapshot();

    rerender(<Component ham={2} />);

    expect(container.firstChild).toMatchSnapshot();

    rerender(<Component ham={2} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
