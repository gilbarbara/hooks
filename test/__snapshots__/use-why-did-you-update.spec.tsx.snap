// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useWhyDidYouUpdate should show the changes 1`] = `<div />`;

exports[`useWhyDidYouUpdate should show the changes 2`] = `
<div>
  {
  "ham": {
    "from": 1,
    "to": 2
  }
}
</div>
`;

exports[`useWhyDidYouUpdate should show the changes 3`] = `<div />`;
